storage/scylla_api.c

```diff
@@ -0,0 +1,616 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+#include "scylla_api.h"
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define SCYLLA_API_LOGGER "SCYLLA_API"
+
+static logger_id_t scylla_api_logger_id;
+
+#define SCYLLA_LOG_ERROR(fmt, ...) \
+  logger_helper_print(scylla_api_logger_id, LOGGER_ERR, "[%s:%d:]" fmt, __func__, __LINE__, __VA_ARGS__)
+#define SCYLLA_LOG_DEBUG(fmt, ...) \
+  logger_helper_print(scylla_api_logger_id, LOGGER_DEBUG, "[%s:%d:]" fmt, __func__, __LINE__, __VA_ARGS__)
+
+void scylla_api_logger_init() { scylla_api_logger_id = logger_helper_enable(SCYLLA_API_LOGGER, LOGGER_DEBUG, true); }
+
+int scylla_api_logger_release() {
+  logger_helper_release(scylla_api_logger_id);
+  if (logger_helper_destroy() != RC_OK) {
+    log_critical(scylla_api_logger_id, "[%s:%d] Destroying logger failed %s.\n", __func__, __LINE__, SCYLLA_API_LOGGER);
+    return EXIT_FAILURE;
+  }
+  return 0;
+}
+
+typedef struct Select_Where_ {
+  cass_byte_t* bundle;
+  cass_byte_t* address;
+} Select_Where;
+
+typedef enum { WITH_BUNDLE = 0, WITH_BUNDLE_AND_ADDRESS, NUM_OF_SELECT_METHOD } Select_Way;
+
+typedef struct Select_Query_ {
+  const char* query;
+} Select_Query;
+
+Transaction* new_Transaction(void) {
+  Transaction* rep = (Transaction*)malloc(sizeof(Transaction));
+  rep->address = malloc(NUM_FLEX_TRITS_ADDRESS);
+  rep->bundle = malloc(NUM_FLEX_TRITS_BUNDLE);
+  rep->transaction = malloc(NUM_FLEX_TRITS_HASH);
+  rep->trunk = malloc(NUM_FLEX_TRITS_TRUNK);
+  rep->branch = malloc(NUM_FLEX_TRITS_BRANCH);
+  rep->message = malloc(NUM_FLEX_TRITS_MESSAGE);
+  return rep;
+}
+void free_Transaction(Transaction** obj) {
+  free((*obj)->address);
+  free((*obj)->bundle);
+  free((*obj)->transaction);
+  free((*obj)->trunk);
+  free((*obj)->branch);
+  free((*obj)->message);
+  free(*obj);
+  return;
+}
+static const Select_Query select_query[NUM_OF_SELECT_METHOD] = {
+    {"SELECT * FROM permanent.bundleTable WHERE bundle = ?"},
+    {"SELECT * FROM permanent.bundleTable WHERE bundle = ? AND address = ?"}};
```

Comment:
Why did you repeat the query string? Can you specify `AND address = ?` instead?

Reply:
The first query is for finding all transactions within a bundle.
The second query is for finding all transactions within an address. But in Scylla, we must specify all the keys before the address as we create the bundle table with primary keys as the following order.
`PRIMARY KEY (bundle, address, transaction)`

With this constraint, we need to get all Bundle hashes to a given address from the Edge table first, then use Bundle hashes and address to get all transaction data from Bundle table.